pkg/controller/controller_instance.go

```diff
@@ -714,21 +714,28 @@ func (c *controller) pollServiceInstance(instance *v1beta1.ServiceInstance) erro
 		switch {
 		case deleting:
 			// For deprovisioning only, we should reattempt even on failure
-			//
-			// TODO(mkibbe): This is actually broken today, as we
-			// will not actually be retrying the deprovision, but
-			// instead constantly re-hitting the "last_operation"
-			// endpoint with the same operation, which will forever
-			// be stuck on failure.
 			msg := "Deprovision call failed: " + description
 			readyCond := newServiceInstanceReadyCondition(v1beta1.ConditionUnknown, errorDeprovisionCalledReason, msg)
 
 			if c.reconciliationRetryDurationExceeded(instance.Status.OperationStartTime) {
 				return c.processServiceInstancePollingFailureRetryTimeout(instance, readyCond)
 			}
 
-			c.processServiceInstanceOperationError(instance, readyCond)
-			return c.continuePollingServiceInstance(instance)
+			setServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionReady, readyCond.Status, readyCond.Reason, readyCond.Message)
+			c.recorder.Event(instance, corev1.EventTypeWarning, errorDeprovisionCalledReason, msg)
+
+			// we must trigger a new deprovision attempt entirely (as
+			// opposed to retrying querying the failed operation
+			// endpoint). Finish polling, and return an error in order
+			// to requeue in the standard instance queue.
+			instance.Status.AsyncOpInProgress = false
+			instance.Status.LastOperation = nil
```

Comment:
How does setting this status trigger the new attempt? As it is this comment can go anywhere in the entire block. I don't understand how the content of the comment relates to the content of the code.

I get finish polling (it's a couple lines below), but how are we returning an error? By setting the ready condition up above? 

Is this doing a pseudo reset of the status, such that when it gets picked back up the logic flows? I feel like there's a smell here if we have to do this reset vs being set elsewhere. And if there is a reset, should it be encapsulated somehow, or the status logic described?

Reply:
The comment was actually stale; I originally was returning an explicit error so that the worker would requeue it, but then changed it to return the result of `finishPollingServiceInstance`, which is usually `nil` and would have its status changes picked up by the watcher and added back to the instance queue as normal.

It's not really a pseudo-reset; it's just clearing the async operation-related fields in the status, so that the resource enters a normal deprovision error state. I made that more explicit now, and I actually like it a lot better this way. PTAL